src/client.rs

```diff
@@ -62,11 +63,64 @@ impl CoinGeckoClient {
     /// let client = CoinGeckoClient::new("https://some.url");
     /// ```
     pub fn new(host: &'static str) -> Self {
-        CoinGeckoClient { host }
+        CoinGeckoClient {
+            host,
+            api_key: None,
+        }
+    }
+
+    /// Creates a new CoinGeckoClient client with a custom host url and API key
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use coingecko::CoinGeckoClient;
+    /// let client = CoinGeckoClient::new_with_api_key("https://some.url", "123456789");
+    /// ```
+    pub fn new_with_api_key(host: &'static str, api_key: &'static str) -> Self {
+        CoinGeckoClient {
+            host,
+            api_key: Some(api_key),
+        }
     }
 
-    async fn get<R: DeserializeOwned>(&self, endpoint: &str) -> Result<R, Error> {
-        reqwest::get(format!("{host}/{ep}", host = self.host, ep = endpoint))
+    /// Gets a URL for the provided endpoint and optional params.
+    /// The endpoint must be prefixed with a /.
+    /// If params are specified, they must be prefixed with a ?.
+    /// If an API key is present, the x_cg_pro_api_key param is added.
+    pub fn get_url(&self, endpoint: &str, params: Option<&str>) -> String {
+        let api_key_param = self
+            .api_key
+            .and_then(|api_key| Some(format!("x_cg_pro_api_key={}", api_key)));
+
+        let params = if let Some(p) = params {
+            // Append the API key param if required
+            format!(
+                "{params}{api_key_param}",
+                params = p,
+                api_key_param = api_key_param.map_or_else(|| "".into(), |key| format!("&{}", key)),
+            )
+        } else {
+            // Supply the API key param if required
+            format!(
+                "{api_key_param}",
+                api_key_param = api_key_param.map_or_else(|| "".into(), |key| format!("?{}", key)),
+            )
+        };
+        format!(
+            "{host}{ep}{params}",
+            host = self.host,
+            ep = endpoint,
+            params = { params }
+        )
```

Comment:
nit: avoid extra format and I see a way to have less code lines that are more understandable. (This might not compile exactly as is)
```suggestion
        let params = match (params, self.api_key) {
          (Some(p), Some(k)) => format!("{p}&x_cg_pro_api_key={k}"),
          (Some(p), None) => p,
          (None, Some(k)) => format!("?x_cg_pro_api_key={k}"),
          (None, None) => "".into(),
        };
        
        format!(
            "{host}{ep}{params}",
            host = self.host,
            ep = endpoint,
            params = { params }
        )
```